flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/TwoInputSelectableStreamTask.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.runtime.execution.Environment;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.streaming.api.operators.TwoInputStreamOperator;
+import org.apache.flink.streaming.runtime.io.StreamTwoInputSelectableProcessor;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+
+import java.util.Collection;
+
+/**
+ * A {@link StreamTask} for executing the {@link TwoInputStreamOperator} which can select the input to
+ * get the next {@link StreamRecord}.
+ */
+@Internal
+public class TwoInputSelectableStreamTask<IN1, IN2, OUT> extends AbstractTwoInputStreamTask<IN1, IN2, OUT> {
+
+	private volatile StreamTwoInputSelectableProcessor<IN1, IN2> inputProcessor;
```

Comment:
What is the purpose of making this field `volatile`?

Reply:
The reason is that the assignment of `inputProcessor` and the call to `inputProcessor.stop()` occur in two different threads (look at the call stacks below), and the execution order is uncertain.

- TaskExecutor.submitTask -> Task.startTaskThread() - -> (**a new thread**) StreamTask.invoke() -> AbstractTwoInputStreamTask.init() -> new StreamTwoInputSelectableProcessor()

- TaskExecutor.cancelTask -> Task.cancelExecution() -> TaskCanceler.start() - -> (**a new thread**) TaskCanceler.run() -> StreamTask.cancel() -> AbstractTwoInputStreamTask.cancelTask() -> StreamTwoInputSelectableProcessor.stop()

Reply:
Yes, it's because of the introduction of this flag. Another comment will explain why "volatile boolean continuousProcessing" is introduced.